src/effects/effectprocessor.h

```diff
@@ -6,100 +7,208 @@
+    // effects be concerned with the ChannelHandle& when process() takes care of giving
+    // it the appropriate ChannelStateHolder?
     virtual void processChannel(const ChannelHandle& handle,
-                                T* channelState,
+                                EffectSpecificState* channelState,
                                 const CSAMPLE* pInput, CSAMPLE* pOutput,
-                                const unsigned int numSamples,
-                                const unsigned int sampleRate,
-                                const EffectProcessor::EnableState enableState,
+                                const mixxx::AudioParameters& bufferParameters,
+                                const EffectEnableState enableState,
                                 const GroupFeatureState& groupFeatures) = 0;
 
-  private:
-    inline T* getOrCreateChannelState(const ChannelHandle& handle) {
-        ChannelStateHolder& holder = m_channelState[handle];
-        if (holder.state == NULL) {
-            holder.state = new T();
+    void process(const ChannelHandle& inputHandle, const ChannelHandle& outputHandle,
+                         const CSAMPLE* pInput, CSAMPLE* pOutput,
+                         const mixxx::AudioParameters& bufferParameters,
+                         const EffectEnableState enableState,
+                         const GroupFeatureState& groupFeatures) final {
+        EffectSpecificState* pState = m_channelStateMatrix[inputHandle][outputHandle];
+        VERIFY_OR_DEBUG_ASSERT(pState != nullptr) {
+            qWarning() << "EffectProcessorImpl::process could not retrieve"
+                          "EffectState for input" << inputHandle
+                       << "and output" << outputHandle
+                       << "EffectState should have been preallocated in the main thread.";
+            pState = createState(bufferParameters);
+            m_channelStateMatrix[inputHandle][outputHandle] = pState;
         }
-        return holder.state;
+        processChannel(inputHandle, pState, pInput, pOutput, bufferParameters,
+                       enableState, groupFeatures);
     }
 
-    ChannelHandleMap<ChannelStateHolder> m_channelState;
+   void initialize(const QSet<ChannelHandleAndGroup>& activeInputChannels,
+            EffectsManager* pEffectsManager,
+            const mixxx::AudioParameters& bufferParameters) final {
+        for (const ChannelHandleAndGroup& inputChannel : activeInputChannels) {
+            //qDebug() << this << "EffectProcessorImpl::initialize allocating EffectStates for input" << inputChannel;
+            ChannelHandleMap<EffectSpecificState*> outputChannelMap;
+            for (const ChannelHandleAndGroup& outputChannel :
+                    pEffectsManager->registeredOutputChannels()) {
+                outputChannelMap.insert(outputChannel.handle(),
+                        createState(bufferParameters));
```

Comment:
The inner loop should move into  createState, since it is duplicated into some places. This will also clarify that the output states are created and destroyed together.

Reply:
good idea

Reply:
Ping!

Reply:
Actually, I'm not sure about this. `createState` is used to create an individual EffectSpecificState in the VERIFY_OR_DEBUG_ASSERT clause of EffectProcessorImpl::process.

Reply:
Ok